torch/csrc/jit/codegen/cuda/fusion_segmenter.cpp

```diff
@@ -856,24 +1078,154 @@ void SegmentCandidateFinder::findSegments() {
     mergeNodes();
   }
 
+  finalMerge();
+
   finalize();
 }
 
+void SegmentCandidateFinder::finalMerge() {
+  AllProducerGroups producer_check(segmented_fusion_.get());
+
+  bool merged_nodes = true;
+  while (merged_nodes) {
+    // Iterate all groups and check if a group
+    //  can merge with one of its consumers
+    for (auto producer_group : groups()) {
+      // Populate consumers and their corresponding consumer edges
+      std::unordered_map<SegmentedGroup*, SegmentedEdge*> consumer_edge_map;
+      std::vector<SegmentedGroup*> all_consumers_of_producer_group;
+      for (auto consumer : producer_group->consumer_edges) {
+        consumer_edge_map.insert({consumer->to, consumer});
+      }
+      // Populate all consumers from the map to avoid duplicate
+      std::transform(
+          consumer_edge_map.begin(),
+          consumer_edge_map.end(),
+          std::back_inserter(all_consumers_of_producer_group),
+          [](auto& it) { return it.first; });
+
+      for (auto consumer : all_consumers_of_producer_group) {
+        if (!producer_check.isConsumerOfAny(
+                consumer, all_consumers_of_producer_group) &&
+            codeGenSupportedMerge(consumer_edge_map.at(consumer))) {
+          to_merge_.emplace_back(producer_group);
+          to_merge_.emplace_back(consumer);
+          producer_group->merged_ = true;
+          producer_group->merge_with_ = consumer;
+          producer_group->merge_through_ = consumer_edge_map.at(consumer);
+          consumer->merged_ = true;
+          consumer->merge_with_ = producer_group;
+          consumer->merge_through_ = producer_group->merge_through_;
+          break;
+        }
+      }
+
+      // Only want to merge one pair at a time so break if found any
+      if (!to_merge_.empty()) {
+        break;
+      }
+    }
+
+    if (to_merge_.empty()) {
+      merged_nodes = false;
+    } else {
+      TORCH_INTERNAL_ASSERT(
+          to_merge_.size() == 2, "merging more than 2 nodes in final iter");
+      auto merged_a = *to_merge_.begin();
+      auto merged_b = merged_a->merge_with_;
+      auto merged_ab = mergeNodes();
+      producer_check.mergeGroups(merged_a, merged_b, merged_ab);
+    }
+  }
+}
+
+void SegmentCandidateFinder::resolveScalarsInGroup(SegmentedGroup* group) {
```

Comment:
@csarofeen current version inserts all the scalar producers to each group before starting merging and doesn't create groups for scalar ops anymore. Will merge this PR if it looks reasonable.